presto-orc/src/main/java/com/facebook/presto/orc/writer/DictionaryColumnWriter.java

```diff
@@ -0,0 +1,491 @@
+
+import static com.facebook.presto.orc.DictionaryCompressionOptimizer.estimateIndexBytesPerValue;
+import static com.facebook.presto.orc.OrcEncoding.DWRF;
+import static com.facebook.presto.orc.metadata.CompressionKind.NONE;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.DATA;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public abstract class DictionaryColumnWriter
+        implements ColumnWriter, DictionaryColumn
+{
+    protected final int column;
+    protected final Type type;
+    protected final CompressionParameters compressionParameters;
+    protected final Optional<DwrfDataEncryptor> dwrfEncryptor;
+    protected final OrcEncoding orcEncoding;
+    protected final MetadataWriter metadataWriter;
+
+    private final LongOutputStream dataStream;
+    private final PresentOutputStream presentStream;
+    private final CompressedMetadataWriter compressedMetadataWriter;
+    private final List<DictionaryRowGroup> rowGroups = new ArrayList<>();
+    private long columnStatisticsRetainedSizeInBytes;
+
+    private IntBigArray rowGroupIndexes;
+    private int rowGroupValueCount;
+    private long rawBytes;
+    private long totalValueCount;
+    private long totalNonNullValueCount;
+    private boolean closed;
+    private boolean inRowGroup;
+    private boolean directEncoded;
+
+    public DictionaryColumnWriter(
+            int column,
+            Type type,
+            CompressionParameters compressionParameters,
+            Optional<DwrfDataEncryptor> dwrfEncryptor,
+            OrcEncoding orcEncoding,
+            MetadataWriter metadataWriter)
+    {
+        checkArgument(column >= 0, "column is negative");
+        this.column = column;
+        this.type = requireNonNull(type, "type is null");
+        this.compressionParameters = requireNonNull(compressionParameters, "compressionParameters is null");
+        this.dwrfEncryptor = requireNonNull(dwrfEncryptor, "dwrfEncryptor is null");
+        this.orcEncoding = requireNonNull(orcEncoding, "orcEncoding is null");
+        LongOutputStream result;
+        if (orcEncoding == DWRF) {
+            result = new LongOutputStreamV1(compressionParameters, dwrfEncryptor, false, DATA);
+        }
+        else {
+            result = new LongOutputStreamV2(compressionParameters, false, DATA);
+        }
+        this.dataStream = result;
+        this.presentStream = new PresentOutputStream(compressionParameters, dwrfEncryptor);
+        this.metadataWriter = requireNonNull(metadataWriter, "metadataWriter is null");
+        this.compressedMetadataWriter = new CompressedMetadataWriter(metadataWriter, compressionParameters, dwrfEncryptor);
+        rowGroupIndexes = new IntBigArray();
+    }
+
+    protected abstract int getInstanceSize();
```

Comment:
This interface together with `getRetainedDictionaryBytes` is weird. Check my other comment below.